### ISSUE TITLE
implicit use of %rax, def of %rax, def of %rdx by IMUL64m not propagated to *.uni

### ISSUE THREAD

**Issue by matsc-at-sics-se** [matsc-at-sics-se/unison#3]
[timing.mir.txt](https://github.com/matsc-at-sics-se/unison/files/1763815/timing.mir.txt)

shell:
```
uni import timing.mir -o timing.uni --target=X86 --maxblocksize=50 --goal=speed --copyremat --lint
allTemporariesUsed: t19 is not used
```

timing.mir:
```
  bb.0 (%ir-block.0, freq 8):
    liveouts:
  
    ADJCALLSTACKDOWN64 0, 0, implicit-def dead %rsp, implicit-def dead %eflags, implicit %rsp
    %0 = MOV32r0 implicit-def dead %eflags
    %1 = LEA64r %stack.0.rusage, 1, _, 0, _
    %edi = COPY %0
    %rsi = COPY %1
    CALL64pcrel32 @getrusage, csr_64, implicit %rsp, implicit %edi, implicit %rsi, implicit-def %rsp, implicit-def %eax
    ADJCALLSTACKUP64 0, 0, implicit-def dead %rsp, implicit-def dead %eflags, implicit %rsp
    %3 = IMUL64rmi32 %stack.0.rusage, 1, _, 0, _, 1000, <0x47c8120> = !{!"unison-memory-partition", i32 0}, implicit-def dead %eflags :: (load 8 from %ir.3)
    %4 = MOV64ri 2361183241434822607
    %rax = COPY %4
    IMUL64m %stack.0.rusage, 1, _, 8, _, <0x47c8150> = !{!"unison-memory-partition", i32 1}, implicit-def dead %rax, implicit-def %rdx, implicit-def dead %eflags, implicit %rax
    %5 = COPY %rdx
    %6 = SHR64ri %5, 63, implicit-def dead %eflags
    %7 = SAR64ri %5, 7, implicit-def dead %eflags
    %8 = ADD64rr %7, killed %6, implicit-def dead %eflags
    %9 = ADD64rr %8, killed %3, implicit-def dead %eflags
    %rax = COPY %9
    RETQ %rax
```

timing.uni, see phoney def t9:rdx by (fun) instead of by IMUL64m

```
b0 (entry, exit, return, freq: 1):
    o0: [t0:rbx,t1:rbp,t2:r12,t3:r13,t4:r14,t5:r15,t6:rsp] <- (in) []
    o1: [] <- ADJCALLSTACKDOWN64 [0,0]
    o2: [t7] <- MOV32r0 []
    o3: [t8] <- LEA64r [%stack.0,1,_,0,_]
    o4: [] <- CALL64pcrel32 [@getrusage]
    o5: [t9:rdx,t10:rax,t11:rcx,t12:rsi,t13:rdi,t14:r8,t15:r9,t16:r10,t17:r11] <- (fun) [t7:edi,t8:rsi] (call: o4)
    o6: [] <- (kill) [t10,t11,t12,t13,t14,t15,t16,t17]
    o7: [] <- ADJCALLSTACKUP64 [0,0]
    o8: [t18] <- IMUL64rmi32 [%stack.0,1,_,0,_,1000] (mem: 0)
    o9: [t19] <- MOV64ri [2361183241434822607]
    o10: [] <- IMUL64m [%stack.0,1,_,8,_] (mem: 1)
    o11: [t20] <- SHR64ri [t9,63]
    o12: [t21] <- SAR64ri [t9,7]
    o13: [t22] <- ADD64rr [t21,t20]
    o14: [t23] <- ADD64rr [t22,t18]
    o15: [] <- RETQ [42]
    o16: [] <- (out) [t0:rbx,t1:rbp,t2:r12,t3:r13,t4:r14,t5:r15,t6:rsp,t23:rax]
```

**Comment by robcasloz** [comment]
Commit 1b2d33505 is also a second step towards fixing this.

**Comment by matsc-at-sics-se** [comment]
extractCallRegs and liftRegs seem to go wrong, I suppose they need to be fixed first:

Output from runTargetTransforms:
```
function: usertime
b0 (entry, exit, return, freq: 8):
    o0: [] <- (in) []
    o1: [] <- ADJCALLSTACKDOWN64 [0,0]
    o2: [t0] <- MOV32r0 []
    o3: [t1] <- LEA64r [%stack.0,1,_,0,_]
    o4: [edi] <- (copy) [t0]
    o5: [rsi] <- (copy) [t1]
    o6: [] <- CALL64pcrel32 [@getrusage]
    o7: [] <- ADJCALLSTACKUP64 [0,0]
    o8: [t3] <- IMUL64rmi32 [%stack.0,1,_,0,_,1000] (mem: 0)
    o9: [t4] <- MOV64ri [2361183241434822607]
    o10: [rax] <- (copy) [t4]
    o11: [rax,rdx] <- IMUL64m [%stack.0,1,_,8,_,rax] (mem: 1)
    o12: [t5] <- (copy) [rdx]
    o13: [t6] <- SHR64ri [t5,63]
    o14: [t7] <- SAR64ri [t5,7]
    o15: [t8] <- ADD64rr [t7,t6]
    o16: [t9] <- ADD64rr [t8,t3]
    o18: [] <- RETQ [42]
    o20: [] <- (out) [t9:rax]
adjacent:
rematerializable:
fixed-frame:
frame:
    %stack.0: offset = 0, size = 144, align = 8
stack-pointer-offset: 0
stack-arg-size: 0
jump-table:
goal: speed
removed-freqs:
source:

```

Output from extractCallRegs. Comments:
o21: def t12:rdx is nonsense. The function has no live output value.
o11: def rax is OK, dead.  def rdx is OK, but needs to be linked to t12.
o11: use rax is OK, comes from o10.  But liftRegs replaces rax by a temp, losing the rax assignment.
```
function: usertime
b0 (entry, exit, return, freq: 8):
    o0: [] <- (in) []
    o1: [] <- ADJCALLSTACKDOWN64 [0,0]
    o2: [t0] <- MOV32r0 []
    o3: [t1] <- LEA64r [%stack.0,1,_,0,_]
    o4: [t10] <- (copy) [t0]
    o5: [t11] <- (copy) [t1]
    o6: [] <- CALL64pcrel32 [@getrusage]
    o21: [t12:rdx] <- (fun) [t10:edi,t11:rsi] (call: o6)
    o7: [] <- ADJCALLSTACKUP64 [0,0]
    o8: [t3] <- IMUL64rmi32 [%stack.0,1,_,0,_,1000] (mem: 0)
    o9: [t4] <- MOV64ri [2361183241434822607]
    o10: [rax] <- (copy) [t4]
    o11: [rax,rdx] <- IMUL64m [%stack.0,1,_,8,_,rax] (mem: 1)
    o12: [t5] <- (copy) [t12]
    o13: [t6] <- SHR64ri [t5,63]
    o14: [t7] <- SAR64ri [t5,7]
    o15: [t8] <- ADD64rr [t7,t6]
    o16: [t9] <- ADD64rr [t8,t3]
    o18: [] <- RETQ [42]
    o20: [] <- (out) [t9:rax]
adjacent:
rematerializable:
fixed-frame:
frame:
    %stack.0: offset = 0, size = 144, align = 8
stack-pointer-offset: 0
stack-arg-size: 0
jump-table:
goal: speed
removed-freqs:
source:
```

Output from liftRegs. The preassignments of t13, t14, t15 were lost:
```
function: usertime
b0 (entry, exit, return, freq: 8):
    o0: [] <- (in) []
    o1: [] <- ADJCALLSTACKDOWN64 [0,0]
    o2: [t0] <- MOV32r0 []
    o3: [t1] <- LEA64r [%stack.0,1,_,0,_]
    o4: [t10] <- (copy) [t0]
    o5: [t11] <- (copy) [t1]
    o6: [] <- CALL64pcrel32 [@getrusage]
    o21: [t12:rdx] <- (fun) [t10:edi,t11:rsi] (call: o6)
    o7: [] <- ADJCALLSTACKUP64 [0,0]
    o8: [t3] <- IMUL64rmi32 [%stack.0,1,_,0,_,1000] (mem: 0)
    o9: [t4] <- MOV64ri [2361183241434822607]
    o10: [t13] <- (copy) [t4]
    o11: [t14,t15] <- IMUL64m [%stack.0,1,_,8,_,t13] (mem: 1)
    o12: [t5] <- (copy) [t12]
    o13: [t6] <- SHR64ri [t5,63]
    o14: [t7] <- SAR64ri [t5,7]
    o15: [t8] <- ADD64rr [t7,t6]
    o16: [t9] <- ADD64rr [t8,t3]
    o18: [] <- RETQ [42]
    o20: [] <- (out) [t9:rax]
adjacent:
rematerializable:
fixed-frame:
frame:
    %stack.0: offset = 0, size = 144, align = 8
stack-pointer-offset: 0
stack-arg-size: 0
jump-table:
goal: speed
removed-freqs:
source:

```

**Comment by matsc-at-sics-se** [comment]
I'm confused. After rebuilding uni, those "affects" and "affected_by" temps are no more in the *.uni:

`    o10: [] <- IMUL64m [t20,t7,%stack.0,1,_] (mem: 1)
`

**Comment by robcasloz** [comment]
Looks like commit 3cf81efa930bb4ae4a69e0cfe8bac03575cbd57d reverted the X86 .hs files (probably due to using an old version of 'specsgen'). Commit d2aa8f16c861fdc2e004f17986c713a6386a7715 fixes that. Please make sure to always build and install the latest 'specsgen' in the repo before running it.

**Comment by matsc-at-sics-se** [comment]
Yes, that was it.  Now it's again:

`    o10: [t20,t21] <- IMUL64m [%stack.0,1,_,8,_,t19] (mem: 1)
`

**Comment by robcasloz** [comment]
The problem with `extractCallRegs` is fixed. The preassigments of `t13`, `t14`, and `t15` by `IMUL64m` need to be reenforced by a target-specific transformation as we discussed offline. Copying the discussion here for reference:

If you look at the output of `uni import` after `liftRegs` you can see:

`[t14,t15] <- IMUL64m [%stack.0,1,_,8,_,t13] (mem: 1)`

What we would like is:

`[t14:rax,t15:rdx] <- IMUL64m [%stack.0,1,_,8,_,t13:rax] (mem: 1)`

For doing this, you can add a target-dependent transformation to `transforms`, at `ImportPostLift` time.

The transformation should go through all operations (`mapToOperation` is your friend), identify the ones implemented by instructions with promoted operands, and insert the appropriate pre-assignments. The key difficulty is to identify which operands result from promoted side-effects, for this you should look at the `promotedRegs` list in `Common.hs` and use `readsSideEffect` and `writesSideEffect` from `Common.hs` much like `promoteImplicitRegs` does in `X86.hs`.

As you pointed out offline, the information generated in `ReadWriteInfo.hs` is incomplete for this task, do you mind creating a separate issue for that?

**Comment by matsc-at-sics-se** [comment]
Fixed by commit 3b46831.
